### PR TITLE
feat: add pin support

### PR DESCRIPTION
**Pull request by skynet2**

### CODE REVIEW

**Review by aholovko — APPROVED** [review]

File: docs/v1/openapi.yaml [hunk 1]
```diff
@@ -1191,6 +1191,9 @@ components:
         tx_id:
           type: string
           description: To be used by Issuer applications for correlation if needed.
+        otp_pin:
+          type: string
+          description: Pre-authorized flow. Generated OTP pin for issuance. Issuer is responsible to deliver that pin to the final client (email,sms,etc).
```
Comment (sudeshrshetty) [inline docs/v1/openapi.yaml:1196]: no need to mention below, explained well in spec 

```Issuer is responsible to deliver that pin to the final client (email,sms,etc)```

rename it from OTP to user_pin matching OIDC spec.

File: pkg/service/oidc4ci/models.go [hunk 2]
```diff
@@ -55,12 +55,12 @@ type TransactionData struct {
 	IssuerAuthCode                     string
 	IssuerToken                        string
 	OpState                            string
-	UserPinRequired                    bool
 	IsPreAuthFlow                      bool
 	PreAuthCode                        string
 	ClaimData                          map[string]interface{}
 	State                              TransactionState
 	WebHookURL                         string
+	OtpPin                             string
```
Comment (sudeshrshetty) [inline pkg/service/oidc4ci/models.go:63]: call it userPIN

File: pkg/service/oidc4ci/oidc4ci_service.go [hunk 3]
```diff
@@ -37,6 +37,11 @@ const (
 
 var logger = log.New("oidc4ci")
 
+type pinGenerator interface {
+	Generate() string
+	Validate(otpKey string, got string) bool
+}
```
Comment (sudeshrshetty) [inline pkg/service/oidc4ci/oidc4ci_service.go:43]: Where do you validate if generated user PIN is bound to a particular transaction?

I would recommend you pass `challenge` to `Generate()` function & later pass that challenge along with user ping for verification.

In VCS open source version, you can ignore challenge if it requires persistence.

File: pkg/service/oidc4ci/pin.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	"fmt"
+	"math/rand"
+	"strings"
+)
+
+const (
+	maxNumber = 10
+)
+
+type PinGenerator struct {
```
Comment (sudeshrshetty) [inline pkg/service/oidc4ci/pin.go:13]: Add Go Doc

File: pkg/service/oidc4ci/oidc4ci_service_initiate_issuance.go [hunk 5]
```diff
@@ -65,11 +65,14 @@ func (s *Service) InitiateIssuance(
...
 		State:                              TransactionStateIssuanceInitiated,
 		WebHookURL:                         profile.WebHook,
 	}
 
+	if req.UserPinRequired {
+		data.OtpPin = s.pinGenerator.Generate()
+	}
+
 	if data.GrantType == "" {
 		data.GrantType = defaultGrantType
```
Comment (sudeshrshetty) [inline pkg/service/oidc4ci/oidc4ci_service_initiate_issuance.go:77]: what is `defaultGrantType ` here?

File: test/bdd/features/oidc4ci_api.feature [hunk 6]
```diff
@@ -33,3 +33,6 @@ Feature: OIDC4CI REST API
 
     When client scans qrcode
     Then client should receive access token for further interactions with vc api
+
+    When client requests credential for claim data with pre-authorize flow
+    Then client receives a valid credential with pre-authorize flow
```
Comment (sudeshrshetty) [inline test/bdd/features/oidc4ci_api.feature:38]: I would recommend adding 2 scenarios - with & without user PIN

File: pkg/service/oidc4ci/oidc4ci_service_initiate_issuance.go (hunk 5 above)
Comment (skynet2) [reply pkg/service/oidc4ci/oidc4ci_service_initiate_issuance.go]: its constant, defaultGrantType    = "authorization_code" @sudeshrshetty

File: component/otp/pin.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright Avast Software. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package otp
```
Comment (sudeshrshetty) [inline component/otp/pin.go:7]: package name is still `otp`

**Pull request merged**